src/mapping/query/function/format_url.rs

```diff
@@ -0,0 +1,235 @@
+        .get(PASSWORD)
+        .and_then(Value::as_option)
+        .map(|v| string(PASSWORD, v))
+        .transpose()?;
+
+    let host = components
+        .get(HOST)
+        .and_then(Value::as_option)
+        .map(|v| string(HOST, v))
+        .transpose()?;
+
+    let port = components
+        .get(PORT)
+        .and_then(Value::as_option)
+        .map(|v| {
+            v.as_i64()
+                .map(|i| i as u16)
+                .ok_or_else(|| invalid(PORT, "map", v.kind()))
+        })
+        .transpose()?;
+
+    let path = components
+        .get(PATH)
+        .map(|v| string(PATH, v))
+        .transpose()?
+        .unwrap_or_else(|| "".to_owned());
+
+    let fragment = components
+        .get(FRAGMENT)
+        .and_then(Value::as_option)
+        .map(|v| string(FRAGMENT, v))
+        .transpose()?;
+
+    let query = components
+        .get(QUERY)
+        .map(|v| v.as_map().ok_or_else(|| invalid(QUERY, "map", v.kind())))
+        .transpose()?
+        .map(|v| {
+            v.iter()
+                .map(|(k, v)| (k, v.to_string_lossy()))
+                .collect::<Vec<_>>()
+        })
+        .unwrap_or_default();
+
+    // This is a really awkward way to create a valid URL from components, but
+    // it doesn't appear the `Url` API offers any better way to handle this.
+    let mut url = Url::parse("https://vector.dev").expect("valid");
+
+    url.set_scheme(&scheme).map_err(|_| unknown(SCHEME))?;
+    url.set_host(host.as_deref()).map_err(|_| unknown(HOST))?;
+    url.set_username(&username).map_err(|_| unknown(USERNAME))?;
+    url.set_password(password.as_deref())
+        .map_err(|_| unknown(PASSWORD))?;
+    url.set_path(&path);
+    url.set_port(port).map_err(|_| unknown(PORT))?;
+    url.set_fragment(fragment.as_deref());
+
+    if !query.is_empty() {
+        url.query_pairs_mut().extend_pairs(query);
+    }
+
+    Ok(url)
+}
+
+fn string(component: &'static str, value: &Value) -> Result<String> {
+    value
+        .as_bytes2()
+        .ok_or_else(|| invalid(component, "string", value.kind()))
+        .map(|b| String::from_utf8_lossy(&b).into_owned())
+}
+
+#[inline]
+fn missing(s: &'static str) -> String {
+    format!("missing url component '{}'", s)
+}
+
+#[inline]
+fn invalid(component: &'static str, expected: &str, got: &str) -> String {
+    format!(
+        "invalid type for url component '{}' (expected {}, got {})",
+        component, expected, got,
+    )
+}
+
+#[inline]
+fn unknown(component: &'static str) -> String {
+    format!("invalid url component '{}'", component)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::collections::BTreeMap;
+
+    macro_rules! map {
+        () => (
+            Value::Map(BTreeMap::new())
+        );
+        ($($k:tt: $v:expr),+ $(,)?) => (
+            Value::from(vec![
+                $(($k.into(), $v.into())),+
+            ].into_iter().collect::<BTreeMap<String, Value>>())
+        );
+    }
```

Comment:
AFAIK we have at least three version of this macro now. I think that would be good to consolidate them somewhere. Not necessary for this PR though.